nucleus/dataset.py

```diff
@@ -1414,11 +1429,76 @@ def items_and_annotation_generator(
+            query=query,
         )
         for data in json_generator:
             for ia in convert_export_payload([data], has_predictions=False):
                 yield ia
 
+    def items_and_annotation_chip_generator(
+        self,
+        chip_size: int,
+        stride_size: int,
+        cache_directory: str,
+        query: Optional[str] = None,
+    ) -> Iterable[Dict[str, str]]:
+        """Provides a generator of chips for all DatasetItems and BoxAnnotations in the dataset.
+
+        Args:
+            chip_size: The size of the image chip
+            stride_size: The distance to move when creating the next image chip.
+              When stride is equal to chip size, there will be no overlap.
+              When stride is equal to half the chip size, there will be 50 percent overlap.
+            cache_directory: The s3 or local directory to store the image and annotations of a chip.
+              s3 directories must be in the format s3://s3-bucket/s3-key
+            query: Structured query compatible with the `Nucleus query language <https://nucleus.scale.com/docs/query-language-reference>`_.
+
+        Returns:
+            Generator where each element is a dict containing the location of the image chip (jpeg) and its annotations (json).
+            ::
+
+                Iterable[{
+                    "image_location": str,
+                    "annotation_location": str
+                }]
+        """
+        json_generator = paginate_generator(
+            client=self._client,
+            endpoint=f"dataset/{self.id}/exportForTrainingPage",
+            result_key=EXPORT_FOR_TRAINING_KEY,
+            page_size=10000,  # max ES page size
+            query=query,
+            chip=True,
+        )
+        for item in json_generator:
+            image = fetch_image(item[ITEM_KEY][PROCESSED_URL_KEY])
+            w, h = image.size
+            annotations = item[BOX_TYPE]
+            item_ref_id = item[ITEM_KEY][REFERENCE_ID_KEY]
+            for x0, y0 in generate_offsets(w, h, chip_size, stride_size):
+                x0, y0 = int(x0), int(y0)
+                ref_id = f"{cache_directory}/{chip_size}/{stride_size}/{item_ref_id}_{x0}_{y0}"
+                chipped_image_loc, chipped_annotation_loc = fetch_chip(ref_id)
+                if chipped_image_loc:
+                    yield {
+                        IMAGE_LOCATION_KEY: chipped_image_loc,
+                        ANNOTATION_LOCATION_KEY: chipped_annotation_loc,
+                    }
+                    continue
+                x1 = x0 + chip_size
+                y1 = y0 + chip_size
+                chipped_image = image.crop((x0, y0, x1, y1))
+                chipped_annotations = chip_annotations(
+                    annotations, x0, y0, x1, y1
+                )
+                chipped_image_loc, chipped_annotation_loc = write_chip(
+                    ref_id, chipped_image, chipped_annotations
+                )
+                yield {
```

Comment:
I would expect this function to generate the image chips and annotations themselves, rather than the locations. Or maybe both.

Reply:
in terms of expectations we do provide in the doc string the output format

the reason I went forward with a location based approach is since this is going to be used on training datasets, users would have to implement this logic themselves (since there's not a great way of storing all of these annotations and images in memory) instead of allowing this to do it all for them.
we would also need to return the pillow image in a dictionary which does not seem very safe

Reply:
Since you are yielding them, they never need to be stored in memory as a whole, and users still get all the benefits of caching if they just call this method with the same cache_directory etc. so they wouldn't lose any of that information.

We have

```py
def items_and_annotation_generator(
        self,
        query: Optional[str] = None,
    ) -> Iterable[Dict[str, Union[DatasetItem, Dict[str, List[Annotation]]]]]:
    ...
```
and it would be best if we also had (pseudocode)
```py
def items_and_annotation_chip_generator(
        self,
        chip_size: int,
        stride_size: int,
        cache_directory: Optional[str] = None,
        query: Optional[str] = None,
    ) -> Iterable[Dict[str, Union[DatasetItem, Dict[str, List[Annotation]]]]]:
    for item in items_and_annotation_generator(query):
        ...
        for x0, y0 in generate_offsets(w, h, chip_size, stride_size):
            try:
                chipped_image, chipped_annotations = fetch_chip(ref_id)
            except NotInCache:
                chipped_image, chipped_annotations = make_chip(ref_id, ...)
                write_chip(ref_id, chipped_image, chipped_annotations)
            yield {...}  # dict formatted like items_and_annotation_generator
```
you already have all the pieces to make that happen, and I think this interface would be easier to use.

Reply:
agree with Robert here

Reply:
we are not able to return items in the way you would like to receive these items @robert-perrotta @bdnorman since once we modify the item, we are no longer returning `DatasetItem`s but instead raw images and its corresponding annotations. with this in mind I believe it's best to stick with the current approach

Reply:
Synced with Vayun, I think this is fine for now and can revisit